src/app/lightning/update-ln-payments.ts

```diff
@@ -0,0 +1,93 @@
+import { LndService } from "@services/lnd"
+import { baseLogger } from "@services/logger"
+import { LnPaymentsRepository } from "@services/mongoose/ln-payments"
+
+export const updateLnPayments = async (): Promise<true | ApplicationError> => {
+  let incompleteLnPayments = await LnPaymentsRepository().listIncomplete()
+  if (incompleteLnPayments instanceof Error) return incompleteLnPayments
+
+  const lndService = LndService()
+  if (lndService instanceof Error) return lndService
+
+  const lndListMethods = [lndService.listSettledPayments, lndService.listFailedPayments]
+  for (const listPaymentsFn of lndListMethods) {
+    incompleteLnPayments = await fetchAndUpdatePayments({
+      incompleteLnPayments,
+      listPaymentsFn,
+    })
+    if (incompleteLnPayments instanceof Error) return incompleteLnPayments
+    if (incompleteLnPayments.length == 0) return true
+  }
+
+  return true
+}
+
+const fetchAndUpdatePayments = async ({
+  incompleteLnPayments,
+  listPaymentsFn,
+}: {
+  incompleteLnPayments: PersistedLnPaymentLookup[]
+  listPaymentsFn: ({
+    after,
+    pubkey,
+  }: ListLnPaymentsArgs) => Promise<ListLnPaymentsResult | LightningServiceError>
+}) => {
+  const processedLnPaymentsHashes: PaymentHash[] | ApplicationError = []
+
+  const lndService = LndService()
+  if (lndService instanceof Error) return lndService
+  const pubkeys = lndService.listPubkeys()
+
+  let lnPayments: LnPaymentLookup[]
+  const endCursors: { [key: Pubkey]: PagingToken | false | undefined } = {}
+  for (const key of pubkeys) endCursors[key] = undefined
+
+  // Breadth-first paginated fetch of payments across multiple lightning instances
```

Comment:
this is the type of comment that make me think we have some important logic here, and maybe `app` is not the right directory? may want to move this either to `domain` or `service` and maybe see if we can test it specifically?

Reply:
I'm not sure we can move this out of the app layer because it involves the interplay between two things from our domain: `LndService` & `LnPaymentsRepository`.

A bit of background, the reason this is "breadth-first" has to do with that we will accumulate payments in lnd for some period (maybe 2 weeks) before deleting, but the cronjob should only ever need to look through about 1 day of data to get all the updates it needs. This loop searches all lnd instances in `TRANSACTION_PER_PAGE` size chunks, instead of searching the entire history of a given lnd before moving on to another. The loop automatically stops querying lnd once all incomplete payments are processed.

Reply:
Ok actually, I think I have an idea for how to move this into the service layer